replica/database.cc

```diff
@@ -1335,6 +1335,33 @@ request_class classify_request(const database_config& _dbcfg) {
 
 } // anonymous namespace
 
+std::optional<db::rate_limiter::can_proceed> database::account_coordinator_write_to_rate_limit(const schema_ptr& s, const dht::token& token) {
+    auto table_limit = s->per_partition_rate_limit_options().get_max_writes_per_second();
+    if (!table_limit || dht::shard_of(*s, token) != this_shard_id() || classify_request(_dbcfg) != request_class::user) {
+        return std::nullopt;
+    }
+    db::per_partition_rate_limit::account_and_enforce info {
+        .participating_groups = 1,
+        .replicas_in_group = 1,
+        .needed_for_cl_in_group = 1,
+    };
+    column_family& cf = find_column_family(s);
+    return _rate_limiter.account_operation(cf.get_rate_limiter_label_for_writes(), dht::token::to_int64(token), *table_limit, info);
+}
+
+std::optional<db::rate_limiter::can_proceed> database::account_coordinator_read_to_rate_limit(const schema_ptr& s, const dht::token& token) {
+    auto table_limit = s->per_partition_rate_limit_options().get_max_reads_per_second();
+    if (!table_limit || dht::shard_of(*s, token) != this_shard_id() || classify_request(_dbcfg) != request_class::user) {
+        return std::nullopt;
+    }
+    db::per_partition_rate_limit::account_and_enforce info {
+        .participating_groups = 1,
+        .replicas_in_group = 1,
+        .needed_for_cl_in_group = 1,
+    };
+    column_family& cf = find_column_family(s);
```

Comment:
Should it not be a method of `table`? Then a caller who already knows the table can avoid a lookup.

Reply:
It needs `_rate_limiter` and `_dbcfg` from `database`, I would have to expose them to the outside callers. Instead I changed this method to take `table&` as an argument.

I hoped to eliminate one call to `find_column_family` this way, but later I deduplicated the code responsible for calculating the right rate limiter info and, unfortunately, the lookup is always there.